File: skills/lasersell/templates/auto_exit_bot.py
"""
LaserSell Auto-Exit Bot Template

Complete, copy-paste-ready bot that:
1. Connects to the LaserSell stream
2. Optionally buys a token
3. Monitors positions and auto-submits exit transactions

Setup:
1. pip install lasersell-sdk solders
2. Set environment variables: LASERSELL_API_KEY, KEYPAIR_PATH
3. Run: python auto_exit_bot.py [optional-mint-to-buy] [optional-amount-sol]
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
from pathlib import Path

from solders.keypair import Keypair

from lasersell_sdk.exit_api import BuildBuyTxRequest, ExitApiClient
from lasersell_sdk.stream.client import StreamClient, StreamConfigure
from lasersell_sdk.stream.session import StreamSession
from lasersell_sdk.tx import SendTargetHeliusSender, send_transaction, sign_unsigned_tx

# --- Configuration ---
API_KEY = os.environ["LASERSELL_API_KEY"]
KEYPAIR_PATH = os.environ.get("KEYPAIR_PATH", "keypair.json")

STRATEGY = {
    "target_profit_pct": 50,
    "stop_loss_pct": 10,
    "trailing_stop_pct": 5,
}
DEADLINE_TIMEOUT_SEC = 0  # 0 = disabled
BUY_SLIPPAGE_BPS = 2_000  # 20%


async def main() -> None:
    keypair = load_keypair(Path(KEYPAIR_PATH))
    wallet_pubkey = str(keypair.pubkey())

    print(f"Wallet: {wallet_pubkey}")
    print(f"Strategy: TP={STRATEGY['target_profit_pct']}% SL={STRATEGY['stop_loss_pct']}% TS={STRATEGY['trailing_stop_pct']}%")

    # 1. Connect stream FIRST
    session = await StreamSession.connect(
        StreamClient(API_KEY),
        StreamConfigure(
            wallet_pubkeys=[wallet_pubkey],
            strategy=STRATEGY,
            deadline_timeout_sec=DEADLINE_TIMEOUT_SEC,
        ),
    )
    print("Stream connected.")

    # 2. Optional: buy a token
    mint_to_buy = sys.argv[1] if len(sys.argv) > 1 else None
    amount_sol = float(sys.argv[2]) if len(sys.argv) > 2 else 0
    if mint_to_buy and amount_sol > 0:
        print(f"Buying {amount_sol} SOL of {mint_to_buy}...")
        exit_client = ExitApiClient.with_api_key(API_KEY)
        buy_response = await exit_client.build_buy_tx(BuildBuyTxRequest(
            mint=mint_to_buy,
            user_pubkey=wallet_pubkey,
            amount=amount_sol,
            slippage_bps=BUY_SLIPPAGE_BPS,
        ))
        signed_buy = sign_unsigned_tx(buy_response.tx, keypair)
        buy_sig = await send_transaction(SendTargetHeliusSender(), signed_buy)
        print(f"Buy submitted: {buy_sig}")

    # 3. Event loop
    print("Monitoring positions...")
    while True:
        event = await session.recv()
        if event is None:
            print("Stream ended.", file=sys.stderr)
            break

        if event.type == "position_opened" and event.handle:
            print(f"[OPEN] {event.handle.mint} (wallet: {event.handle.wallet_pubkey})")

        elif event.type == "pnl_update":
            pnl = event.message.get("pnl_pct", "?")
            pid = event.message.get("position_id", "?")
            print(f"[PNL] position={pid} pnl={pnl}%")

        elif event.type == "exit_signal_with_tx" and event.message.get("type") == "exit_signal_with_tx":
            try:
                signed = sign_unsigned_tx(event.message["unsigned_tx_b64"], keypair)
                sig = await send_transaction(SendTargetHeliusSender(), signed)
                print(f"[EXIT] {event.message['mint']} signature={sig}")
            except Exception as err:
                print(f"[EXIT FAILED] {event.message.get('mint', '?')}: {err}", file=sys.stderr)

        elif event.type == "position_closed":
            print(f"[CLOSED] position={event.message.get('position_id', '?')}")


def load_keypair(path: Path) -> Keypair:
    raw = path.read_text(encoding="utf-8")
    parsed = json.loads(raw)
    if not isinstance(parsed, list):
        raise ValueError("Keypair file must be a JSON array of numbers")
    return Keypair.from_bytes(bytes(int(v) for v in parsed))


if __name__ == "__main__":
    asyncio.run(main())

File: skills/lasersell/templates/auto-exit-bot.ts
/**
 * LaserSell Auto-Exit Bot Template
 *
 * Complete, copy-paste-ready bot that:
 * 1. Connects to the LaserSell stream
 * 2. Optionally buys a token
 * 3. Monitors positions and auto-submits exit transactions
 *
 * Setup:
 * 1. npm install @lasersell/lasersell-sdk @solana/web3.js
 * 2. Set environment variables: LASERSELL_API_KEY, KEYPAIR_PATH
 * 3. Run: npx tsx auto-exit-bot.ts [optional-mint-to-buy] [optional-amount-sol]
 */
import { readFile } from "node:fs/promises";
import { Keypair } from "@solana/web3.js";
import {
  ExitApiClient,
  StreamClient,
  StreamSession,
  StrategyConfigBuilder,
  proveOwnership,
  signUnsignedTx,
  sendTransaction,
  sendTargetHeliusSender,
} from "@lasersell/lasersell-sdk";

// --- Configuration ---
const API_KEY = process.env.LASERSELL_API_KEY!;
const KEYPAIR_PATH = process.env.KEYPAIR_PATH || "keypair.json";

const STRATEGY = new StrategyConfigBuilder()
  .targetProfitPct(50)
  .stopLossPct(10)
  .trailingStopPct(5)
  .liquidityGuard(true)
  .build();
const DEADLINE_TIMEOUT_SEC = 0; // 0 = disabled
const BUY_SLIPPAGE_BPS = 2_000; // 20%

// --- Main ---
async function main(): Promise<void> {
  const keypair = await loadKeypair(KEYPAIR_PATH);
  const walletPubkey = keypair.publicKey.toBase58();

  console.log(`Wallet: ${walletPubkey}`);
  console.log(`Strategy: ${JSON.stringify(STRATEGY)}`);

  // 1. Prove wallet ownership (local Ed25519 signature, no network call)
  const proof = proveOwnership(keypair);

  // 2. Register wallet (required before stream connection)
  const apiClient = ExitApiClient.withApiKey(API_KEY);
  await apiClient.registerWallet(proof);

  // 3. Connect stream with registered wallet
  const session = await new StreamClient(API_KEY).connectWithWallets(
    [proof],
    STRATEGY,
    DEADLINE_TIMEOUT_SEC,
  );
  console.log("Stream connected.");

  // 4. Optional: buy a token
  const mintToBuy = process.argv[2];
  const amountSol = parseFloat(process.argv[3] || "0");
  if (mintToBuy && amountSol > 0) {
    console.log(`Buying ${amountSol} SOL of ${mintToBuy}...`);
    const exitClient = ExitApiClient.withApiKey(API_KEY);
    const buyResponse = await exitClient.buildBuyTx({
      mint: mintToBuy,
      user_pubkey: walletPubkey,
      amount: amountSol,
      slippage_bps: BUY_SLIPPAGE_BPS,
      input: "SOL",
    });
    const signedBuy = signUnsignedTx(buyResponse.tx, keypair);
    const buySig = await sendTransaction(sendTargetHeliusSender(), signedBuy);
    console.log(`Buy submitted: ${buySig}`);
  }

  // 5. Event loop
  console.log("Monitoring positions...");
  while (true) {
    const event = await session.recv();
    if (event === null) {
      console.error("Stream ended.");
      break;
    }

    switch (event.type) {
      case "position_opened":
        console.log(`[OPEN] ${event.handle.mint} (wallet: ${event.handle.wallet_pubkey})`);
        break;

      case "pnl_update":
        if ("pnl_pct" in event.message) {
          console.log(`[PNL] position=${event.message.position_id} pnl=${event.message.pnl_pct}%`);
        }
        break;

      case "exit_signal_with_tx":
        if (event.message.type === "exit_signal_with_tx") {
          try {
            const signed = signUnsignedTx(event.message.unsigned_tx_b64, keypair);
            const sig = await sendTransaction(sendTargetHeliusSender(), signed);
            console.log(`[EXIT] ${event.message.mint} signature=${sig}`);
          } catch (err) {
            console.error(`[EXIT FAILED] ${event.message.mint}: ${err}`);
          }
        }
        break;

      case "position_closed":
        console.log(`[CLOSED] position=${event.message.position_id}`);
        break;

      case "liquidity_snapshot":
        // Tier 1+ only. Uncomment to log.
        // console.log(`[LIQUIDITY] position=${event.message.position_id} trend=${event.message.trend}`);
        break;
    }
  }
}

async function loadKeypair(path: string): Promise<Keypair> {
  const raw = await readFile(path, "utf8");
  const parsed = JSON.parse(raw);
  if (!Array.isArray(parsed)) throw new Error("Keypair file must be a JSON array of numbers");
  return Keypair.fromSecretKey(Uint8Array.from(parsed.map(Number)));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
